File: backend/src/main/java/com/sisenco/backend/model/Report.java
package com.sisenco.backend.model;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * @author Dusan
 * @date 9/6/2026
 */

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "reports")
public class Report {
    @Id
    private String id;

    private String userId; // Team Member ගේ ID එක
    private String projectId; // Project/Category ID එක

    private Date weekStartDate;
    private Date weekEndDate;

    private ReportStatus status = ReportStatus.DRAFT; // Default status

    private List<TaskRecord> tasksCompleted = new ArrayList<>();
    private List<String> tasksPlannedForNextWeek = new ArrayList<>();

    private List<IssueRecord> blockers = new ArrayList<>();
    private List<AchievementRecord> achievements = new ArrayList<>();

    // Optional Fields
    private String notes;

    // Review & Workflow Data
    private int currentVersion = 1;
    private String latestManagerComment;

    private Date createdAt = new Date();
    private Date updatedAt = new Date();
}

File: frontend/src/service/report.ts
import api from "./api";
import { PaginatedResponse } from "./manager";

export interface TaskRecord {
  taskName: string;
  priority: string;
  plannedVsActualPercentage: string;
  taskStatus: string;
  plannedVsSpentTime: string;
  output: string;
}

export interface IssueRecord {
  description: string;
  isKeyIssue: boolean;
}

export interface AchievementRecord {
  description: string;
  isKeyAchievement: boolean;
}

export interface ReportRequestDto {
  projectId: string;
  weekStartDate: string;
  weekEndDate: string;
  tasksCompleted: TaskRecord[];
  tasksPlannedForNextWeek: string[];
  blockers: IssueRecord[];
  achievements: AchievementRecord[];
  notes: string;
  isSubmit: boolean; 
}

export interface Report extends Omit<ReportRequestDto, 'isSubmit'> {
  id: string;
  userId: string;
  status: string;
  currentVersion: number;
  latestManagerComment?: string;
  createdAt: string;
  updatedAt: string;
}

export const createReportAPI = async (data: ReportRequestDto) => {
  const response = await api.post('/reports', data);
  return response.data.data as Report;
};

export const updateReportAPI = async (id: string, data: ReportRequestDto) => {
  const response = await api.put(`/reports/${id}`, data);
  return response.data.data as Report;
};

export const getMyReportsAPI = async (page: number = 0, size: number = 10) => {
  const response = await api.get(`/reports/my-reports?page=${page}&size=${size}`);
  return response.data.data as PaginatedResponse<Report>;
};

export const getReportByIdAPI = async (id: string) => {
  const response = await api.get(`/reports/${id}`);
  return response.data.data as Report;
};